some constitutive_law flags added and cleaned

josep-m-carbonell opened the pull request:
- The constitutive law flags were used in a wrong manner because some more local flags were needed
- This new flags have been added (check them)
- A very general flag (not descriptive) has been removed
- Some changes in Dam and PoroMechanics applications have been applied and they need a review

josep-m-carbonell committed:
clean constitutive_law flags

useless flags are removed and neede flags are added

RiccardoRossi reviewed (changes requested):
Please update the documentation. Or at least comment on the suggested correct usage.

It is not clear to me...

[EDIT:] just to clarify, i am asking to document/discuss the behaviour change for the base class, not for all the files

ipouplana commented:
Indeed, as @josep-m-carbonell suggests, some new ConstitutiveLaw flags were necessary, e.g. MECHANICAL_RESPONSE_ONLY, THERMAL_RESPONSE_ONLY, NON_LOCAL_RESOLUTION. However, I think it would be better if we could replace the latter flag by LOCAL_RESOLUTION or LOCAL_RESOLUTION_ONLY.

Until now, when working with non-local damage models, we computed first the local quantities through the methods InitializeNonLinearIteration and FinalizeNonLinearIteration by setting the flag ISOCHORIC_TENSOR_ONLY (because there were no better ones available), and then we calculated their non-local counterpart through the rest of methods (FinalizeSolutionStep,CalculateLocalSystem,CalculateRightHandSide, CalculateLeftHandSide...) without setting any special flag. The think is, using the NON_LOCAL_RESOLUTION flag would require changing all those methods instead of just one (because in fact FinalizeNonLinearIteration calls InitializeNonLinearIteration).

If @josep-m-carbonell has no inconveniences I can replace the NON_LOCAL_RESOLUTION flag by the LOCAL_RESOLUTION one.

pooyan-dadvand commented:
First of all I should declare that I'm not an expert in constitutive laws. 

Saying that I'm ok with adding new local flags to constitutive laws base class. Nevertheless deleting an existing one is much more delicated and need a proper advertisement before accepting this pull request.

Meanwhile, looking to the code, it become very complex with several layers of ifs for options. This means that each constitutive law is dealing with too many different situations. Is there any way to have a base class for each and seperate derived one for each option combination? (This called change cases with class hierarchy)

Finally there are several noalias are missing which would improve the performance. (I can point them in the code if is interested)

ipouplana commented:
If @pooyan-dadvand can point out the noalias missing I would appreciate it. However, I don't actually see the possibility of distributing the different ConstitutiveLaw options among derived classes. There are already a lot of ConstitutiveLaw classes and that would make it even worse...

josep-m-carbonell commented:
@pooyan-dadvand  @RiccardoRossi 

Discussing this flags with @ipouplana, LOCAL_RESOLUTION, is the default resolution and NON_LOCAL_RESOLUTION does not solve the problems because a lot of particularization is needed. Instead, a INITIALIZE_MATERIAL_RESPONSE would be better for previous computations before the standard material response calculation. Maybe a NON_LOCAL_LAW flag is needed for the features.

Yes @pooyan-dadvand , the TOTAL_TENSOR flag can be kept if it is needed, but it defines the default behaviour, it can be confussing. I guess that other structure or hierarchy is possible, but I don't know if it will solve the complexity of the situations, but adding some extra files. 

About the nonalias, you are right. Only the flags were reviewed, I you point at them we will make the corrections.

@RiccardoRossi  I will add some comments in the flag definitions. The discussion for improvements is open.

Thank you

ipouplana commented:
Thanks @josep-m-carbonell, if there is no inconvenience this afternoon I will update some files in this branch substituting the NON_LOCAL_RESOLUTION flag by a new INITIALIZE_MATERIAL_RESPONSE flag.

josep-m-carbonell committed:
delete files from missing application

RiccardoRossi commented:
Hi @ipouplana,
            there is currently one documentation of the constitutive law (in the old wiki). I did it quite a lot of time ago, and most probably it is outdated.

What i am asking is that if new flags are passed, than it should be explained what they are for. a good way for doing this would be to migrate the old documentation to the new wiki, and to update it with a description of what the new flags are expected to do.
if the change is accepted that would set the new rule to be followed ... however we definitely need to reach an agreement before modifying the base class.

having said this, and since we discuss the flags, i am not clear at all about the expected behaviour for the flag COMPUTE_STRAIN. This would also be a good moment to clarify that

I also think that the strain energy should be computed differently from what is done now, as the current behaviour implies storing the strain energy within the class which is something that should be avoided as much as possible.

pooyan-dadvand commented:
Just to clarify that I'm not against the cleanup. if the flag is not needed anymore it is better to be cleaned sooner than later!

pooyan-dadvand commented on applications/DamApplication/custom_constitutive/thermal_linear_elastic_3D_law.cpp:
```diff
@@ -64,41 +64,31 @@ void ThermalLinearElastic3DLaw::CalculateMaterialResponseKirchhoff (Parameters&
-			Vector ThermalStrainVector;
-			this->CalculateThermalStrain(ThermalStrainVector,ElasticVariables,Temperature);
-
-			StressVector = prod(ConstitutiveMatrix,(StrainVector - ThermalStrainVector));
```
`noalias(StressVector)`

pooyan-dadvand commented on applications/DamApplication/custom_constitutive/thermal_linear_elastic_3D_law.cpp:
```diff
@@ -64,41 +64,31 @@ void ThermalLinearElastic3DLaw::CalculateMaterialResponseKirchhoff (Parameters&
+	this->CalculateThermalStrain(ThermalStrainVector,ElasticVariables,Temperature);
 
-        StressVector = prod(ConstitutiveMatrix,(StrainVector - ThermalStrainVector));
+	StressVector = prod(ConstitutiveMatrix,(StrainVector - ThermalStrainVector));
```
same...

pooyan-dadvand commented on applications/DamApplication/custom_constitutive/thermal_linear_elastic_3D_law.cpp:
```diff
@@ -107,13 +97,35 @@ void ThermalLinearElastic3DLaw::CalculateMaterialResponseKirchhoff (Parameters&
         this->CalculateThermalStrain(StrainVector,ElasticVariables,Temperature);
 
         StressVector = prod(ConstitutiveMatrix,StrainVector);
```
also

pooyan-dadvand commented on applications/DamApplication/custom_constitutive/thermal_linear_elastic_3D_law.cpp:
```diff
@@ -107,13 +97,35 @@ void ThermalLinearElastic3DLaw::CalculateMaterialResponseKirchhoff (Parameters&
+        Vector ThermalStrainVector;
+        this->CalculateThermalStrain(ThermalStrainVector,ElasticVariables,Temperature);
+
+        StressVector = prod(ConstitutiveMatrix,(StrainVector - ThermalStrainVector));
```
also

RiccardoRossi replied:
also here my advice is to generate a tmp explicitly for the difference of vectors

    array_1d<double,6> tmp = StrainVector - ThermalStrainVector
    noalias(StressVector) = prod(ConstitutiveMAtrix, tmp);

according to the ublas documentation, in this context creating explicitly the tmp is better than working with the expression resulting of StrainVector - ThermalStrainVector

ipouplana replied:
Thanks @RiccardoRossi, I'll take this into account.

ipouplana replied:
Sorry @RiccardoRossi, now that I look carefully, it is not possible to define an array_1d<double,6> for the temporal vector because the method CalculateMaterialResponseCauchy is inside a 3D ConstitutiveLaw that is also used for 2D plane stress/plane strain laws. 

In order to avoid creating a derived method for each law, we could define a dynamic vector of Voigt size, but... would there be any improvement with respect to the current code?

ipouplana replied:
In the end I've done the following:

`Vector tmp(StrainVector.size());`
`noalias(tmp) = StrainVector - ThermalStrainVector;`
`noalias(StressVector) = prod(ConstitutiveMatrix,tmp);`

pooyan-dadvand commented on applications/DamApplication/custom_constitutive/thermal_linear_elastic_3D_law_nodal.cpp:
```diff
@@ -65,56 +65,69 @@ void ThermalLinearElastic3DLawNodal::CalculateMaterialResponseKirchhoff (Paramet
+ 
+	this->CalculateLinearElasticMatrix( ConstitutiveMatrix, YoungModulus, PoissonCoefficient );
+	
+	StressVector = prod(ConstitutiveMatrix,StrainVector);
```
`noalias(StressVector)`

pooyan-dadvand commented on applications/DamApplication/custom_constitutive/thermal_linear_elastic_3D_law_nodal.cpp:
```diff
@@ -65,56 +65,69 @@ void ThermalLinearElastic3DLawNodal::CalculateMaterialResponseKirchhoff (Paramet
+	
+	this->CalculateThermalStrain(StrainVector,ElasticVariables,Temperature);
+	
+	StressVector = prod(ConstitutiveMatrix,StrainVector);
```
It seems that a search for `StressVector` and changing them for `noalias(StressVector)` whould be necessary, so I stop here.

pooyan-dadvand reviewed (commented):
I have tried to point some noalias but seems that most of them can be found easily using a search by StressVector

josep-m-carbonell commented:
@RiccardoRossi, I agree about the COMPUTE_STRAIN, it was there to ask the constitutive law to compute it if the strain was not supplied. Now I think that this is not the matter of this flag, it is more related of obtaining certain strain information coming from the constitutive law that the element can not supply. It depends also on where things are stored, for example the initial and historical strains.

About the computation of the strain energy, the evaluation of a method or the computation of stresses is needed. What would be the different way to obtain it. Any proposal is welcome.

Thank you again

pooyan-dadvand commented:
@josep-m-carbonell and @ipouplana , Honestly I understand that you don't want to have more classes. But in a long term vision I suggest to find a solution avoiding many ifs. Considering that in a future if we want to add some new physics to the CLs, all becomes even more complex. 

One option would be to have some Internal:: classes and few CLs that use them (may be via template)

ipouplana commented:
Thanks @pooyan-dadvand. In a few minutes I am going to update some files in this branch, and I will add the missing noalias.

RiccardoRossi commented:
@josep-m-carbonell 
regarding the strain_energy calculation, as i understand the problem is that the "Calculate" function does ot have sufficient data to compute it, so you must store it. If there was a Calculate function accepting the parameters needed for actual calculations (more or less the same as CalculateMaterialREsponse + VARIABLE + output space) that would do right?
I am suggesting this since we have very many CLs (one per gauss point) so any saving in memory per gauss point is very relevant.

concerning the COMPUTE_STRAIN flag, what about changing it to STRAIN_PROVIDED_AS_INPUT or something of the sort?

josep-m-carbonell commented:
@RiccardoRossi . I agree that we need to save as memory as possible, to add a "CalculateMethod" similar to CalculateMaterialResponse to obtain other variables can fit the purpose. 

About the flag COMPUTE_STRAIN, I think it is not needed anymore to set that the strain was provided as input. Lets say that this is something expected (the strain to be provided if needed), we can avoid to set this information and solve the current cases where we used this flag. 
If it does not have any other use type, maybe is better to take this flag away instead of replace it for another one with a better name....

josep-m-carbonell commented:
@pooyan-dadvand  You are right about the number of "ifs", in this particular case a different structure would be better. This is not so usual. We usually have an "if" to compute ConstituveMatrix and Stress, or Stress only.  Can you give us an example for the use of a template structure to avoid these "ifs" ?

Something that has to be extended also, is the InitializeMaterialResponse method similar to the FinalizeMaterialResponse, to be in consonance with the INITIALIZE_MATERIAL_RESPONSE flag that @ipouplana added. It can be used internally but it would not set from the element.

ipouplana committed:
Substituted NON_LOCAL_RESOLUTION by INITIALIZE_MATERIAL_RESPONSE.

ipouplana reviewed (approved):
I just updated the ConstitutiveLaws affected by the flags' change. I have compiled it and tested in the PoromechanicsAppliacation and everything seems to work fine. 

However, I would ask @lorenzogracia to check whether the tests in the DamApplication also run properly before merging this PR.

josep-m-carbonell committed:
add initializematerial response methods

lorenzogracia commented:
hi all,
thanks @pooyan-dadvand  for the noalias suggestions. I have run the test examples of the dam application and everything works fine. Nonetheless, I have been realized that REFERENCE_TEMPERATURE has been moved to the core so I take advantage  and I update the change. 

I have no objections.

lorenzogracia committed:
Updating dam thermomechanical solver

ipouplana committed:
Updated dam mpi thermo mechanical solver

pooyan-dadvand commented:
@josep-m-carbonell after our converstation yesterday I was thinking about how to decompose so many combinations in Constitutive Laws and still I don't have a clear solution. As we talked yesterday, here we have two sever constrains: Performance and memory occupation. These constraines prevent us from using usuall patterns like composite or decorator in their generic form.

templates can make the componed pattern faster but avoiding data duplication is quite tricky. The idea is to have a composite pattern by template which takes i.e two CLs (like ElastoPlastic and Thermal) and combines them. Obviously there are two major problems here: 

- The CLs should be composeable (at least for specific sequence of method calls)
- To avoid data duplication, we should pass the data in an object (the compiste CL itself is an option) to all the methods (violating the encapsulation for efficiency reason)

Finally, I need to have more insight of the CLs to be more precise on this. But I think you and @RiccardoRossi can find a solution

RiccardoRossi commented:
@ipouplana, 
Has the doc been ported?

ipouplana commented:
Hi @RiccardoRossi , I just uploaded a page in the Wiki on how to use the constitutive law class with a brief explanation of the new flags (https://github.com/KratosMultiphysics/Kratos/wiki/How-to-use-the-Constitutive-Law-class).

Obviously this is just a first version that can be modified according to the structure that the constitutive law should have in the future...

josep-m-carbonell committed:
Merge branch 'master' into feature-clean-claw-flags

josep-m-carbonell committed:
Merge branch 'master' into feature-clean-claw-flags

josep-m-carbonell committed:
corrections for master compatibility

josep-m-carbonell committed:
corrections in a conflicting file

a hidden merge destroyed the file of this pull request, an automatic resolution of github turn it to a mess.

josep-m-carbonell committed:
Merge branch 'master' into feature-clean-claw-flags

josep-m-carbonell committed:
Merge branch 'master' into feature-clean-claw-flags

josep-m-carbonell committed:
set methods to calculate an ask for a variable

RiccardoRossi commented:
@ipouplana, first of all thx for updating the documentation.

there are a few points i would like to discuss:
1. the base class constitutive law should have nothing to do specifically with a damage model, be it local or non-local. 
2. How would u call the INITIALIZE_MATERIAL_RESPONSE/FINALIZE_MATERIAL_RESPONSE flags of the CL? what is the expected behaviour if an element does not call them? Recall that the design principle is that any constitutive law should be usable by any element, or expose the impossibility so that an error is thrown in the Check method.
3. the COMPUTE_STRAIN_ENERGY method currently requires storing the energy being computed within the CL. To my undestanding, the compute methods added by @josep-m-carbonell would allow getting read of such limitation (btw, my first impression of them is very positive). If this is correct, **can we get rid of this flag**?
4.  i am not very much in favour of the flag     INCREMENTAL_STRAIN_MEASURE. I appreciate why this is done, but a CL that exposes such behaviour is absolutely incompatible with elements that use the complete strain measure. Also the total F is needed anyhow for push/pull back operation, so that one anyway needs to store both Fn and DF. **Do we REALLY need this option**? If so, how do we write a check so that it fails if coupling with an incompatible element?
5. Let's now imagine that i have a standard TL element which does not know at all if the problem to be solved is thermomechanical or not. If i use a temperature dependent CL i would expect it to include the thermal response by default. how is this compatible with the flags MECHANICAL_RESPONSE_ONLY/THERMAL_RESPONSE_ONLY?. shall we assume that such flags default to false? - not that this is a "documentation" question, essentially i am suggesting to document that if the user does not provide them then total response should be expected. How will we know if an element provides such feature (for example at the moment of doing the check)? 
6. same comment as above applies to  ISOCHORIC_TENSOR_ONLY/VOLUMETRIC_TENSOR_ONLY. Can we default those to false? (nad document that). 
7. This is a suggestion: i believe that the usage of the flag COMPUTE_STRAIN is not clear. I would suggest to rename this as USE_ELEMENT_PROVIDED_STRAIN (setting this to false would imply that the CL will compute internally the StrainVector). A kludge is that the element should allocate the StrainVector in the parameters in any case. (a better naming for the flag would also be very welcome)
8. this is mostly a TODO for myself and for @vryy. The "old" CL interface has been deprecated for years now. I think that the legacy application "structural_application" (lower case, not the modern StructuralMechanicsApplication) should implement a legacy_cl_wrapper and derive all of the constitutive laws from that, so that the "new" interface maps to the old and deprecated one. Then all the cls using the old interface could be used using the new interface, and we could finally remove the old interface from the API
9. can anyone put some example of usage of the pull_back/push_forward capabilities of the base class in the documentation?

RiccardoRossi commented:
@ipouplana **PING**! (:-D)

ipouplana commented:
Hi @RiccardoRossi , 
I saw your response but, right now, I have neither the knowledge nor the time to answer to all those questions (I need to finish my thesis for September).
Since they are important topics affecting the Kratos core, I believe that a meeting with all the interested people would be a better format to discuss this...
Regards

RiccardoRossi commented:
Hi @ipouplana , we can definitely have a meeting to simplify the discussion, however it is difficult to include people outside of cimne (i am thinking in particular about @vryy and @AndreasWinterstein).
So, if someone has comments, even on just some of the points i rise it would be a starter, as i guess we could reach a easy agreement on at least some points.

to start easy here a few concrete proposals, which would be very easy to implement (at least the first 2)

1. **Default to false** the flags
     MECHANICAL_RESPONSE_ONLY
     THERMAL_RESPONSE_ONLY
     ISOCHORIC_TENSOR_ONLY
     VOLUMETRIC_TENSOR_ONLY
     **which is to say that most CLs will not even look into the value of such flags**
2.  change COMPUTE_STRAIN --> USE_ELEMENT_PROVIDED_STRAIN
3. remove the flag COMPUTE_STRAIN_ENERGY and replace it by the call to the new functions added
4. remove the flag INCREMENTAL_STRAIN_MEASURE

@KratosMultiphysics/structural-mechanics @ipouplana @lorenzogracia @josep-m-carbonell @jcotela @pooyan-dadvand @AndreasWinterstein @loumalouomega @marandra @AFranci is there anyone against this? (if so, please specify which of the changes you are against)

ipouplana commented:
Thanks @RiccardoRossi for the simplification ^_^'
The changes you propose seem fine to me. Perhaps removing the flag COMPUTE_STRAIN_ENERGY could be a bit bothersome for some applications (e.g. I was using it in the PoromechanicsApplication to distinguish a particular case for the interface elements... probably I can change this flag by another one...)
Regards

josep-m-carbonell commented:
Hi @ipouplana @RiccardoRossi , I will have more freedom next week to discuss some of your questions. Maybe to put that questions and suggestions in an issue, will also be more visibe for everybody. 

Just to answer quickly, first 3 points, for me are OK, however in point 2 (I think is ok to change the name)  are still some indefinitions because you can supply an strain but you must also say what strain is supplied.
The other point I would like to discuss is about rates and increments (point 5), I think a flag is needed to not duplicate the storage of varialbes in Elements a ConstitutiveLaws, we can think in a good way to do it.

KR

RiccardoRossi commented:
@josep-m-carbonell,

your point on the user-provided strain is fair.
I think we should limit that option exclusively to the small deformation case (or to the rate form). 
My point is that if someone is working in large deformations than i believe it is ok to require providing the F instead. However in small deformations computing the hack on the basis of the strain is a sort of hack. 

for the moment, when i can i'll change the wiki with the documentation about point 1. 
Point 3 requires a deeper action, so for the moment i will not do anything about it.


finally i have another question: right now in the constitutive law there are the methods

     FinalizeMaterialResponse*
     FinalizeSolutionStep

what is the intended usage of that? Should they both be called in the FinalizeSolutionStep of the element?

RiccardoRossi commented:
@josep-m-carbonell 

would u agree about defining a function "HasFeatureFlag" in the base class, something like

     bool HasFeatureFlag(Flag feature_flag)
     {
         if(feature_flag == MECHANICAL_RESPONSE_ONLY) return true;
         if(feature_flag == THERMAL_RESPONSE_ONLY  ) return false;
         if(feature_flag == ISOCHORIC_TENSOR_ONLY    ) return false;
         if(feature_flag == VOLUMETRIC_TENSOR_ONLY ) return false;
     }

so to check if a given behaviour is supported? 
the point then would be that a CL using the last 3 flags would have to reimplement such function and expose its capabilities

josep-m-carbonell commented:
Hello @RiccardoRossi , for me it is ok. However I will not recommend  to implement the constitutive laws with the use of these flags if there are other possible structures to get the requiered information from the constitutive law. For legacy, and to keep certain coding structures compatibility, it is ok.

Kind regards

RiccardoRossi commented:
i definitely agree with you, however i am not clear on what would be the "correct" mechanism here.

for example what would be the reccomended way to get  the volumetric response or the thermal_response? (note that i don't know the usage ofthe flags either, simply i am not sure of what to reccomend in this use case)

pooyan-dadvand commented:
@RiccardoRossi Using Flag a better implementation would be:

``` C++
bool Is(Flag TheFeatures)
 {
     constexpr features = MECHANICAL_RESPONSE_ONLY; // you can use & to add other features 
     return features.Is(TheFeatures)
 }
```

RiccardoRossi commented:
@pooyan-dadvand @josep-m-carbonell 

i think that my proposal is not needed, and it is best implemented in the GetLawFeatures. 
(which is to say that i suggest forgetting my last post)

on a side note i sensibly improved the documentation in the wiki, and i would appreciate if someone can take a look into that (mostly to verify i am not saying something stupid)

ipouplana commented:
Hi @RiccardoRossi , I read your changes of the documentation in the wiki and I think everything seems fine according to what is being proposed here. I just had a couple of doubts/comments. 
First, when you explain the "internal flags":
" These are flags that are useful for computation internally to the CL. they should NOT be used by the element. "
What do you mean by "they should not be used by the element"? Then, how are they supposed to be used? Aren't flags a way of communication between the element and the CL?
Second, when talking about the getvalue/setvalue methods, you wrote twice the same example of the GetValue method, but I suppose you wanted to write the GetValue and then the SetValue method... If it is like this, I can modify it.

RiccardoRossi commented:
Hi @ipouplana 
after also speaking with @josep-m-carbonell my understanding is that those flags should be intended as internal tools of the CL class. That is they do not belong to the API, and are just to be used within the internal implementation. 
If i don't get it wrong, the idea is that If you need to finalizeTheMaterial response then you should call the corresponding function. does this constitute a problem for you?

regarding the SetValue i fixed it, thx for the hint!

ipouplana commented:
Ok @RiccardoRossi , I misunderstood the meaning of the internal flags. In any case, I agree with the idea of calling specific methods, e.g. FinalizeMaterialResponseCauchy, instead of using those flags in the elements.

josep-m-carbonell commented:
@ipouplana  @RiccardoRossi, about he  INCREMENTAL_STRAIN_MEASURE flag. What is your proposal for passing a strain rate to the law. The strain measure can be a non symmetric tensor. For example dF/dt.

ipouplana commented:
@josep-m-carbonell if the tendency of the new CLs is to minimize the usage of flags, perhaps we could create a new method named CalculateIncrementalResponse or something similar...

RiccardoRossi committed:
API BREAKING - changing COMPUTE_STRAIN -> USE_ELEMENT_PROVIDED_STRAIN

RiccardoRossi committed:
API BREAKING - changing COMPUTE_STRAIN -> USE_ELEMENT_PROVIDED_STRAIN

RiccardoRossi commented:
@josep-m-carbonell then let's leave the INCREMENTAL flag...

RiccardoRossi commented:
i just committed to the branch the change COMPUTE_STRAIN -> USE_ELEMENT_PROVIDED_STRAIN

josep-m-carbonell committed:
Merge branch 'master' into feature-clean-claw-flags

josep-m-carbonell committed:
adapting files with the last change in flags

josep-m-carbonell committed:
Merge branch 'master' into feature-clean-claw-flags

RiccardoRossi commented:
Just to tell that i don't dislike the idea of having a 

CalculateMaterialResponse

and a 

CalculateMaterialResponse_Incremental

this way one could throw an error in the CalculateMaterialResponse in the case the _Incremental is to be employed

josep-m-carbonell commented:
So you  mean that in the Incremental case you will expect incremental deformation Epsilon and incremental F parameters instead of the total ones? And in InitializeMaterialResponse and FinalizeMaterialResponse and CalculateValue ..., the incremental methods would be added too?

josep-m-carbonell commented:
Thank you @RiccardoRossi, we will keep on improving the constitutive law
